lib/sync/index.js

```diff
@@ -78,27 +78,59 @@ function invoke(dataset_id, params, callback) {
 
 // Initialise cloud data sync service for specified dataset.
 function init(dataset_id, options, cb) {
-  syncUtil.doLog(dataset_id, 'info', 'init');
-  datasets.init(dataset_id, options);
-  return cb && cb(null, {});
+  syncUtil.doLog(dataset_id, 'info', 'init sync for dataset');
+  start(function(err){
+    if (err) {
+      return cb(err);
+    }
+    datasets.init(dataset_id, options);
+    syncStorage.updateManyDatasetClients({datasetId: dataset_id}, {stopped: false}, cb);
+  });
 }
 
 // Stop cloud data sync for the specified dataset_id.
 function stop(dataset_id, cb) {
-  syncUtil.doLog(dataset_id, 'info', 'stop');
-  
-  // TODO
-
-  cb(null, {});
+  if (!syncStarted) {
+    return cb();
+  }
+  syncUtil.doLog(dataset_id, 'info', 'stop sync for dataset');
+  syncStorage.updateManyDatasetClients({datasetId: dataset_id}, {stopped: true}, cb);
 }
 
-// Stop cloud data sync service for ALL datasets.
+// Stop cloud data sync service for ALL datasets and reset.
+// This should really only used by tests.
 function stopAll(cb) {
-  syncUtil.doLog(syncUtil.SYNC_LOGGER, 'info', 'stopAll');
-  
-  // TODO
-
-  cb(null, {});
+  //sync is not started yet, but connect could be called already. In this case, just reset a few things
+  if (!syncStarted) {
+    mongoDbClient = null;
+    redisClient = null;
+    metricsClient = null;
+    return cb();
+  }
+  syncUtil.doLog(syncUtil.SYNC_LOGGER, 'info', 'stopAll syncs');
+  async.parallel([
+    async.apply(syncStorage.updateManyDatasetClients, {}, {stopped: true}),
+    async.apply(syncWorker.stop.bind(syncWorker)),
+    async.apply(ackWorker.stop.bind(ackWorker)),
+    async.apply(pendingWorker.stop.bind(pendingWorker)),
+    async.apply(syncScheduler.stop.bind(syncScheduler))
+  ], function(err){
+    if (err) {
+      syncUtil.doLog(syncUtil.SYNC_LOGGER, 'error', 'Failed to stop sync due to error : ' + util.inspect(err));
+      return cb(err);
+    }
+    mongoDbClient = null;
+    redisClient = null;
+    metricsClient = null;
+    ackQueue = null;
+    pendingQueue = null;
+    syncQueue = null;
+    ackWorker = null;
+    pendingWorker = null;
+    syncWorker = null;
+    syncStarted = false;
```

Comment:
yeah, good point. I see no reason why we can't do that. Will make index.js a little bit cleaner as well.